influx-reports/influx.py

```diff
@@ -0,0 +1,248 @@
+        self.output_stream = open(self.output_path, 'wb')
+        self.output = StreamResultToBytes(self.output_stream)
+        self.output.startTestRun()
+        return self.output
+
+    def __exit__(self, *args, **kwargs):
+        self.output.stopTestRun()
+        self.output_stream.close()
+
+    def status(self, *args, **kwargs):
+        self.output.status(*args, **kwargs)
+
+
+def return_time(client, query, delta_seconds=0):
+    """ From an InfluxDB query, fetch
+    the first point time, and return a
+    python time object. Shift it from
+    a few seconds (delta_seconds) if necessary.
+    """
+    timestamp_query = client.query(query)
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(
+            fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as alle:
+        raise InfluxTimestampParseException(
+            "Error parsing a timestamp from influx: {}".format(alle))
+    return fpt
+
+
+def calculate_measurement_uptime(client, measurement, job_reference,
+                                 measurement_period_seconds,
+                                 resolution=60):
+    """ For a certain build (job_reference), estimate the amount
+    of seconds a component (column of table) was up.
+    For that, we slice the job by resolution.
+    If I have no data in the slice, I assume downtime. If I have
+    at least one data per slice, I assume up, even if degraded.
+    The count of these positive events multiplied by resolution
+    is an average idea of the uptime.
+    """
+    per_field_downtime_seconds = defaultdict(int)
+    per_field_downtime_percent = {}
+
+    query = (
+        "select max(/.*_status/) from {measurement} "
+        "where time < now() and job_reference='{job_reference}' "
+        "group by time({resolution}s) fill(-1)"
+    ).format(measurement=measurement,
+             job_reference=job_reference,
+             resolution=resolution)
+
+    all_data = client.query(query)
+    for time_slice in all_data.get_points():
+        for element in time_slice:
+            if element != 'time' and time_slice[element] == 1:
+                key = element.replace('max_', '')
+                per_field_downtime_seconds[key] += resolution
+                per_field_downtime_percent[key] = \
+                    (float(per_field_downtime_seconds[key]) /
+                        measurement_period_seconds) * 100
+
+    # total_seconds is the measurement period multiplied by the number of
+    # series. This is used to create an average percent for the measurement.
+    total_seconds = len(per_field_downtime_seconds) \
+        * measurement_period_seconds
+    total_downtime = sum(per_field_downtime_seconds.values())
+    measurement_downtime_seconds = total_downtime \
+        / len(per_field_downtime_seconds)
+    measurement_downtime_percent = \
```

Comment:
The numbers I'm getting when I generate a report look a bit off, for example heat was down for over 7000% of the time! I'm a misunderstanding the numbers?
```---
measurement_downtime_percent:
  maas_cinder: 14.661654135338345
  maas_glance: 44.3609022556391
  maas_heat: 7233.082706766918
  maas_horizon: 39.097744360902254
  maas_keystone: 43.609022556390975
  maas_neutron: 40.100250626566414
  maas_nova: 2739.6616541353387
```